server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java

```diff
@@ -522,6 +525,59 @@ public void writeFileChunk(StoreFileMetaData fileMetaData, long position, BytesR
         }
     }
 
+    @Override
+    public CompletableFuture<Void> writeFileChunk(StoreFileMetaData fileMetaData, long position, BytesReference content,
+                                                  boolean lastChunk, int totalTranslogOps) throws IOException {
+        state().getTranslog().totalOperations(totalTranslogOps);
+        final FileChunkWriter writer = fileChunkWriters.computeIfAbsent(fileMetaData.name(), name -> new FileChunkWriter());
+        writer.writeChunk(new FileChunk(fileMetaData, content, position, lastChunk));
+        return CompletableFuture.completedFuture(null);
+    }
+
+    private static final class FileChunk {
+        final StoreFileMetaData md;
+        final BytesReference content;
+        final long position;
+        final boolean lastChunk;
+        FileChunk(StoreFileMetaData md, BytesReference content, long position, boolean lastChunk) {
+            this.md = md;
+            this.content = content;
+            this.position = position;
+            this.lastChunk = lastChunk;
+        }
+    }
+
+    private final class FileChunkWriter {
+        // chunks can be delivered out of order, we need to buffer chunks if there's a gap between them.
+        final PriorityQueue<FileChunk> pendingChunks = new PriorityQueue<>(Comparator.comparing(fc -> fc.position));
+        long lastPosition = 0;
+
+        void writeChunk(FileChunk newChunk) throws IOException {
+            synchronized (this) {
+                pendingChunks.add(newChunk);
```

Comment:
If a message is dropped or delayed for quite a bit, we possibly keep on adding and adding to this queue, which can become quite big. I wonder if we need more control on the sender side to prevent sending chunks until some prior chunks have been handled.

Reply:
+1 Good call. I will work on this.